File: software/android-sdk-probe/app/src/main/java/io/xstarrevival/sdkprobe/MainActivity.java
package io.xstarrevival.sdkprobe;

import android.os.Bundle;
import android.widget.ScrollView;
import android.widget.TextView;

import com.autel.common.CallbackWithOneParam;
import com.autel.common.battery.BatteryState;
import com.autel.common.error.AutelError;
import com.autel.common.product.AutelProductType;
import com.autel.sdk.Autel;
import com.autel.sdk.ProductConnectListener;
import com.autel.sdk.battery.XStarBattery;
import com.autel.sdk.product.BaseProduct;
import com.autel.sdk.product.XStarPremiumAircraft;

import android.support.v7.app.AppCompatActivity;

import java.util.Arrays;

/**
 * Deliberately read-only engineering probe.
 *
 * This class exposes no arm, motor, takeoff, landing, mission, calibration,
 * firmware, parameter-write, or battery-write APIs.
 */
public final class MainActivity extends AppCompatActivity {
    private TextView output;
    private XStarBattery battery;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        output = new TextView(this);
        output.setTextSize(16f);
        output.setPadding(32, 32, 32, 32);

        ScrollView scroll = new ScrollView(this);
        scroll.addView(output);
        setContentView(scroll);

        line("X-STAR OFFICIAL SDK PROBE");
        line("READ-ONLY / PROPS-OFF BENCH MODE");
        line("");
        line("SDK auth: " + app().getAuthStatus());
        line("Waiting for product connection…");

        Autel.setProductConnectListener(new ProductConnectListener() {
            @Override
            public void productConnected(BaseProduct product) {
                runOnUiThread(() -> inspectProduct(product));
            }

            @Override
            public void productDisconnected() {
                clearBatteryListener();
                runOnUiThread(() -> line("Product disconnected"));
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        line("SDK auth: " + app().getAuthStatus());
    }

    @Override
    protected void onDestroy() {
        clearBatteryListener();
        super.onDestroy();
    }

    private void inspectProduct(BaseProduct product) {
        if (product == null) {
            line("Connected callback returned null product");
            return;
        }

        AutelProductType type = product.getType();
        line("");
        line("PRODUCT CONNECTED");
        line("Type: " + type);
        line("Class: " + product.getClass().getName());

        if (type != AutelProductType.PREMIUM || !(product instanceof XStarPremiumAircraft)) {
            line("Read-only battery probe currently enables only PREMIUM.");
            line("No control methods will be attempted.");
            return;
        }

        XStarPremiumAircraft aircraft = (XStarPremiumAircraft) product;
        battery = (XStarBattery) aircraft.getBattery();

        if (battery == null) {
            line("Battery interface: unavailable");
            return;
        }

        line("Battery interface: available");
        subscribeBatteryState();
        readBatterySnapshot();
    }

    private void subscribeBatteryState() {
        battery.setBatteryStateListener(new CallbackWithOneParam<BatteryState>() {
            @Override
            public void onSuccess(BatteryState state) {
                runOnUiThread(() -> line("Battery state: " + String.valueOf(state)));
            }

            @Override
            public void onFailure(AutelError error) {
                runOnUiThread(() -> line("Battery listener error: " + XStarProbeApplication.safeError(error)));
            }
        });
    }

    private void readBatterySnapshot() {
        battery.getVoltageCells(new CallbackWithOneParam<int[]>() {
            @Override
            public void onSuccess(int[] cells) {
                runOnUiThread(() -> line("Cell voltages: " + Arrays.toString(cells)));
            }

            @Override
            public void onFailure(AutelError error) {
                runOnUiThread(() -> line("Cell voltage error: " + XStarProbeApplication.safeError(error)));
            }
        });

        battery.getVoltage(floatCallback("Pack voltage"));
        battery.getCurrent(floatCallback("Current"));
        battery.getTemperature(floatCallback("Temperature"));
        battery.getDesignCapacity(floatCallback("Design capacity"));
        battery.getCapacity(floatCallback("Capacity"));

        battery.getRemainingPercent(new CallbackWithOneParam<Integer>() {
            @Override
            public void onSuccess(Integer value) {
                runOnUiThread(() -> line("Remaining percent: " + value));
            }

            @Override
            public void onFailure(AutelError error) {
                runOnUiThread(() -> line("Remaining percent error: " + XStarProbeApplication.safeError(error)));
            }
        });
    }

    private CallbackWithOneParam<Float> floatCallback(final String label) {
        return new CallbackWithOneParam<Float>() {
            @Override
            public void onSuccess(Float value) {
                runOnUiThread(() -> line(label + ": " + value));
            }

            @Override
            public void onFailure(AutelError error) {
                runOnUiThread(() -> line(label + " error: " + XStarProbeApplication.safeError(error)));
            }
        };
    }

    private void clearBatteryListener() {
        if (battery != null) {
            try {
                battery.setBatteryStateListener(null);
            } catch (Throwable ignored) {
                // Diagnostic app must not crash during teardown.
            }
            battery = null;
        }
    }

    private XStarProbeApplication app() {
        return (XStarProbeApplication) getApplication();
    }

    private void line(String value) {
        if (output == null) return;
        output.append(value + "\n");
    }
}

File: software/android-sdk-probe/app/src/main/java/io/xstarrevival/sdkprobe/MainActivity.kt
package io.xstarrevival.sdkprobe

import android.os.Bundle
import android.widget.ScrollView
import android.widget.TextView
import android.support.v7.app.AppCompatActivity
import io.xstarrevival.autelsdk.OfficialAutelSdkBridge
import io.xstarrevival.core.adapter.AutelSdkPlatformAdapter
import io.xstarrevival.core.model.ConnectionState
import io.xstarrevival.core.model.XStarState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

/** Props-off engineering UI backed only by the receive-only official SDK bridge. */
class MainActivity : AppCompatActivity() {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private lateinit var output: TextView
    private var platform: AutelSdkPlatformAdapter? = null
    private var usbInventory: List<String> = emptyList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        output = TextView(this).apply {
            textSize = 16f
            setPadding(32, 32, 32, 32)
        }
        setContentView(ScrollView(this).apply { addView(output) })
        usbInventory = UsbInventory.describe(this)

        val appKey = BuildConfig.AUTEL_APP_KEY
        if (appKey.isBlank()) {
            renderNoKey()
            return
        }

        val bridge = OfficialAutelSdkBridge(applicationContext, appKey)
        val sdkPlatform = AutelSdkPlatformAdapter(scope, bridge)
        platform = sdkPlatform

        scope.launch {
            sdkPlatform.state.collectLatest(::renderState)
        }
        scope.launch {
            runCatching { sdkPlatform.connect() }
        }
    }

    override fun onDestroy() {
        platform?.let { sdkPlatform -> runBlocking { sdkPlatform.disconnect() } }
        scope.cancel()
        super.onDestroy()
    }

    private fun renderNoKey() {
        output.text = buildString {
            appendLine("X-STAR OFFICIAL SDK PROBE")
            appendLine("READ-ONLY / PROPS-OFF BENCH MODE")
            appendLine()
            appendLine("SDK auth: NO APP KEY")
            appendLine("Set AUTEL_APP_KEY locally; it is never committed or logged.")
            appendUsbInventory()
        }
    }

    private fun renderState(state: XStarState) {
        output.text = buildString {
            appendLine("X-STAR OFFICIAL SDK PROBE")
            appendLine("READ-ONLY / PROPS-OFF BENCH MODE")
            appendLine()
            appendLine("CONNECTION")
            appendLine(connectionText(state.connection))
            appendLine("Product: ${state.aircraft.productName ?: "—"}")
            appendLine()
            appendLine("BATTERY")
            appendLine("Remaining: ${state.battery.percent?.let { "$it%" } ?: "—"}")
            appendLine("Pack: ${state.battery.packVoltageV?.let { "%.3f V".format(it) } ?: "—"}")
            appendLine("Current: ${state.battery.currentA?.let { "%.3f A".format(it) } ?: "—"}")
            appendLine("Temperature: ${state.battery.temperatureC?.let { "%.1f °C".format(it) } ?: "—"}")
            state.battery.cells.forEach { cell ->
                appendLine("Cell ${cell.index}: ${cell.voltageV?.let { "%.3f V".format(it) } ?: "—"}")
            }
            appendLine()
            appendLine("FLIGHT / SENSORS")
            appendLine("GPS: ${state.navigation.gpsFix ?: "—"} · sats ${state.navigation.satellites ?: "—"}")
            appendLine("Altitude: ${state.navigation.altitudeM?.let { "%.2f m".format(it) } ?: "—"}")
            appendLine("Ultrasonic: ${state.navigation.ultrasonicHeightM?.let { "%.3f m".format(it) } ?: state.navigation.ultrasonicHeightRaw?.let { "%.3f raw SDK units".format(it) } ?: "—"}")
            appendLine("Attitude: roll ${formatAngle(state.attitude.rollDeg)}, pitch ${formatAngle(state.attitude.pitchDeg)}, yaw ${formatAngle(state.attitude.yawDeg)}")
            appendLine("Mode: ${state.aircraft.flightMode ?: "—"}")
            appendLine()
            appendLine("REMOTE / IMAGE LINK")
            appendLine("Remote: ${state.remote.connected ?: "—"}")
            appendLine("RC signal: ${state.remote.signalPercent?.let { "$it%" } ?: "—"}")
            appendLine("RC battery: ${state.remote.batteryPercent?.let { "$it%" } ?: "—"}")
            appendLine("Image signal: ${state.remote.imageSignalPercent?.let { "$it%" } ?: "—"}")
            appendLine("USB enabled: ${state.imageLink.usbEnabled ?: "—"}")
            appendLine("RF: ${state.imageLink.rfFrequencyHz?.let { "%.3f MHz".format(it / 1_000_000.0) } ?: "—"} · raw ${state.imageLink.rfSignalValue ?: "—"}")
            appendLine()
            appendLine("GIMBAL / CAMERA / VIDEO")
            appendLine("Gimbal: ${state.gimbal.status ?: "—"} · ${formatAngle(state.gimbal.pitchDeg)}")
            appendLine("Camera: ${state.camera.connected ?: "—"} · ${state.camera.mode ?: "—"}")
            appendLine("Exposure: ${state.camera.exposureMode ?: "—"} · ISO ${state.camera.iso ?: "—"} · ${state.camera.shutter ?: "—"}")
            appendLine("H.264: ${if (state.camera.video.receiving) "RECEIVING" else "NO STREAM"} · ${state.camera.video.framesReceived} frames")
            appendLine()
            appendLine("COMPONENT VERSIONS")
            state.aircraft.componentVersions.toSortedMap().forEach { (name, value) -> appendLine("$name: $value") }
            appendLine()
            appendLine("DIAGNOSTICS")
            state.diagnostics.counters.toSortedMap().forEach { (name, value) -> appendLine("$name: $value") }
            state.diagnostics.notes.takeLast(6).forEach(::appendLine)
            appendUsbInventory()
        }
    }

    private fun StringBuilder.appendUsbInventory() {
        appendLine()
        appendLine("PASSIVE USB INVENTORY")
        usbInventory.forEach(::appendLine)
    }

    private fun connectionText(connection: ConnectionState): String = when (connection) {
        ConnectionState.Disconnected -> "Disconnected"
        ConnectionState.Discovering -> "Discovering"
        is ConnectionState.Connecting -> "Connecting: ${connection.stage}"
        is ConnectionState.Connected -> "Connected: ${connection.transport}"
        is ConnectionState.Failed -> "Failed: ${connection.stage} · ${connection.reason}"
    }

    private fun formatAngle(value: Double?): String = value?.let { "%.1f°".format(it) } ?: "—"
}
